sympy/functions/special/spherical_harmonics.py

```diff
@@ -189,16 +192,22 @@ def _eval_rewrite_as_sin(self, n, m, theta, phi, **kwargs):
     def _eval_rewrite_as_cos(self, n, m, theta, phi, **kwargs):
         # This method can be expensive due to extensive use of simplification!
         from sympy.simplify import simplify, trigsimp
-        # TODO: Make sure n \in N
-        # TODO: Assert |m| <= n ortherwise we should return 0
```

Comment:
I interpret these as different requirements.
1) n should be in S.Naturals, if so, if `Abs(m) < n`. So while the comparison will break for complex n, I do not think one should return 0 if n is not in S.Naturals (but one may have to change something other than the comment to see...).

Reply:
But n should return 0 when `Abs(m) > n` ? Not sure of what should be returned when n is not in S.Naturals.

Reply:
If I understand it correctly, in general, the rewrite can only happen for these cases. So unless it can be rewritten one should return self (or possibly None, not sure how it works in detail).

So, the previous code just applies the rewrite, although it should only be applied if the requirements hold. (The problem is that when you rewrite like this you can possibly evaluate cases that really are not valid if they were evaluated using the original function.) But this is how I interpret it.

Reply:
None.

Compare to this from factorial. Only if the argument is integer and non-negative, the rewrite is valid:

Reply:
I've made a commit and rewrote the code but I'm not sure of the  `_eval_conjugate` one as of now. For that I'm checking if `phi` and `theta` are numbers and if satisfied then I'm checking the whether `phi` and `theta` are in `S.Reals` and returning `None` if they aren't in `S.Reals`.

Reply:
```
>>> theta = Symbol("theta")
>>> phi = Symbol("phi")
```
I've added real = True  for the above lines in the docstring but for some reason theta and phi aren't being evaluated as Reals.